httpx/concurrency/asyncio/compat.py

```diff
@@ -0,0 +1,138 @@
+        ...
+
+    async def read(self, n: int = -1) -> bytes:
+        ...
+
+    async def start_tls(
+        self,
+        sslContext: ssl.SSLContext,
+        *,
+        server_hostname: typing.Optional[str] = None,
+        ssl_handshake_timeout: typing.Optional[float] = None,
+    ) -> None:
+        ...
+
+    def write(self, data: bytes) -> typing.Awaitable[None]:
+        ...
+
+
+async def connect_compat(*args: typing.Any, **kwargs: typing.Any) -> Stream:
+    if sys.version_info >= (3, 8):
+        return await asyncio.connect(*args, **kwargs)
+    else:
+        reader, writer = await asyncio.open_connection(*args, **kwargs)
+        return StreamCompat(reader, writer)
+
+
+class StreamCompat:
+    """
+    Thin wrapper around asyncio.StreamReader/StreamWriter to make them look and
+    behave similarly to an asyncio.Stream.
+    """
+
+    def __init__(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
+        self.reader = reader
+        self.writer = writer
+
+    def at_eof(self) -> bool:
+        return self.reader.at_eof()
+
+    def close(self) -> typing.Awaitable[None]:
+        self.writer.close()
+        return _OptionalAwait(self.wait_closed)
+
+    async def drain(self) -> None:
+        await self.writer.drain()
+
+    def get_extra_info(self, name: str, default: typing.Any = None) -> typing.Any:
+        return self.writer.get_extra_info(name, default)
+
+    async def read(self, n: int = -1) -> bytes:
+        return await self.reader.read(n)
+
+    async def start_tls(
+        self,
+        sslContext: ssl.SSLContext,
+        *,
+        server_hostname: typing.Optional[str] = None,
+        ssl_handshake_timeout: typing.Optional[float] = None,
+    ) -> None:
+        if not sys.version_info >= (3, 7):  # pragma: no cover
+            raise NotImplementedError(
+                "asyncio.AbstractEventLoop.start_tls() is only available in Python 3.7+"
+            )
+        else:
+            # This code is in an else branch to appease mypy on Python < 3.7
+
+            reader = asyncio.StreamReader()
+            protocol = asyncio.StreamReaderProtocol(reader)
+            transport = self.writer.transport
+
+            loop = asyncio.get_event_loop()
+            loop_start_tls = loop.start_tls  # type: ignore
+            tls_transport = await loop_start_tls(
+                transport=transport,
+                protocol=protocol,
+                sslcontext=sslContext,
+                server_hostname=server_hostname,
+                ssl_handshake_timeout=ssl_handshake_timeout,
+            )
+
+            reader.set_transport(tls_transport)
+            self.reader = reader
+            self.writer = asyncio.StreamWriter(
+                transport=tls_transport, protocol=protocol, reader=reader, loop=loop
+            )
+
+    def write(self, data: bytes) -> typing.Awaitable[None]:
+        self.writer.write(data)
+        return _OptionalAwait(self.drain)
```

Comment:
`await`-ing on the `drain` call is not _exactly_ the same as how things work in Python 3.8 when you `await` on `Stream.write()` it first tries a fast-path to flush data and then falls back to a full drain.

Reply:
Apparently [drain() was deprecated in 3.8 too](https://docs.python.org/3.8/library/asyncio-stream.html#asyncio.Stream.drain), and it is now recommended to `await stream.write()`. Should we switch to this latter API? This means we'd need to refactor `AsyncioBackend.write()` a bit, and potentially use a manual buffering mechanism for `.write_no_block()`. (There's an example of this in the trio backend: #276.)

Reply:
I was hoping this could tie into #341 quite nicely maybe?

Reply:
Yes, actually it does. :)